contracts/solver-registry/src/lib.rs

```diff
@@ -0,0 +1,116 @@
+use dcap_qvl::{verify, QuoteCollateralV3};
+use hex::{decode, encode};
+use near_sdk::{
+    assert_one_yocto,
+    env::{self, block_timestamp},
+    ext_contract, log, near, require,
+    store::{IterableMap, IterableSet, Vector},
+    AccountId, PanicOnDefault, Promise, PublicKey,
+};
+
+use crate::pool::*;
+use crate::types::*;
+
+mod collateral;
+mod events;
+mod pool;
+mod token_receiver;
+mod types;
+mod upgrade;
+mod view;
+
+#[near(serializers = [json, borsh])]
+#[derive(Clone)]
+pub struct Worker {
+    pool_id: u32,
+    checksum: String,
+    codehash: String,
+}
+
+#[near(contract_state)]
+#[derive(PanicOnDefault)]
+pub struct Contract {
+    owner_id: AccountId,
+    intents_contract_id: AccountId,
+    pools: Vector<Pool>,
+    approved_codehashes: IterableSet<String>,
+    worker_by_account_id: IterableMap<AccountId, Worker>,
+}
+
+#[allow(dead_code)]
+#[ext_contract(ext_intents_vault)]
+trait IntentsVaultContract {
+    fn add_public_key(intents_contract_id: AccountId, public_key: PublicKey);
+}
+
+#[near]
+impl Contract {
+    #[init]
+    #[private]
+    pub fn new(owner_id: AccountId, intents_contract_id: AccountId) -> Self {
+        Self {
+            owner_id,
+            intents_contract_id,
+            pools: Vector::new(Prefix::Pools),
+            approved_codehashes: IterableSet::new(Prefix::ApprovedCodeHashes),
+            worker_by_account_id: IterableMap::new(Prefix::WorkerByAccountId),
+        }
+    }
+
+    pub fn register_worker(
+        &mut self,
+        pool_id: u32,
+        quote_hex: String,
+        collateral: String,
+        checksum: String,
+        tcb_info: String,
+    ) -> Promise {
+        require!(self.has_pool(pool_id), "Pool not found");
+
+        let collateral = collateral::get_collateral(collateral);
+        let quote = decode(quote_hex).unwrap();
+        let now = block_timestamp() / 1000000000;
+        let result = verify::verify(&quote, &collateral, now).expect("report is not verified");
+        let rtmr3 = encode(result.report.as_td10().unwrap().rt_mr3.to_vec());
+        let codehash = collateral::verify_codehash(tcb_info, rtmr3);
+
+        // only allow workers with approved code hashes to register
+        require!(self.approved_codehashes.contains(&codehash));
+
+        log!("verify result: {:?}", result);
+
+        // TODO: verify predecessor implicit account is derived from this public key
```

Comment:
you also need to check that this key is generated from within the TEE. This should be part of the quote verification. The RTMR3 hash should be extended with the public key, i.e, `rtmr3 = hash(rmtr3 || public_key)`

Reply:
Please remove this comment then

Reply:
This TODO is to verify the predecessor account's address (implicit account) matches with the public key. This can be optional.

Let me confirm and may add this enhancement later.